bubbles/plugins/__base__.py

```diff
@@ -0,0 +1,184 @@
+    _subclasses = []
+
+    # These should be filled out in subclasses
+    start_at: timedelta
+    interval: timedelta
+
+    def __init__(self):
+        Thread.__init__(self, name=self.__class__.__name__)
+        self.name = self.__class__.__name__
+        self.first_run = True
+        self.stopped = Event()
+
+    @abstractmethod
+    def job(self) -> Any:
+        """
+        This is where the meat of the backgrounded, periodically executed
+        job is defined.
+        """
+        ...
+
+    def run(self):
+        while not self.stopped.wait((self.start_at if self.first_run else self.interval).total_seconds()):
+            self.job()
+            self.first_run = False
+
+    def stop(self):
+        self.stopped.set()
+        self.join()
+
+    def __init_subclass__(cls, **kwargs):
+        """
+        Hook that fires when a class is loaded which inherits from this
+        class. We use this hook to track the subclasses (which are added
+        to the registry) in a way that they can be removed later if need
+        be. Such as providing an automated test harness. Hint hint.
+        """
+        super().__init_subclass__(**kwargs)
+
+        # We override this instead of using __subclasses__()
+        # because __subclasses__() cannot be overridden to
+        # remove a class. Even if it's dynamically generated.
+        # This messes with our automated testing strategy and
+        # could become an issue further down the road if we
+        # have an item we want to remove from the registry
+        #
+        # See: https://stackoverflow.com/a/43057166
+        cls._subclasses.append(cls)
+
+
+class BaseRegistry(ABC):
+    """
+    A mixin that defines an interface common to all registries we manage
+
+    Functionality:
+      - Logger instance
+      - Context manager (e.g., `with MyRegistry() as r:`)
+      - Subclass loader (invoked automatically by context manager)
+    """
+
+    _log: logging.Logger
+
+    @property
+    def log(self) -> logging.Logger:
+        if not hasattr(self, '_log'):
+            self._log = logging.getLogger(__name__)
+
+        return self._log
+
+    @abstractmethod
+    def load():
+        ...
+
+    def __enter__(self):
+        self.load()
+        return self
+
+    def __exit__(self, *_):
+        ...
+
+
+class EventLoop(BaseRegistry):  # pragma: no cover
+    jobs: Set[BasePeriodicJob] = set([])
+
+    def __init__(self, *_, **kwargs):
+        self.reddit: Reddit = kwargs['reddit']
+
+    def __enter__(self):
+        self.load()
+        self.start()
+
+        return self
+
+    def __exit__(self):
+        self.stop()
+
+    def start(self):
+        for job in self.jobs:
+            job.start()
+
+    def stop(self):
+        for job in self.jobs:
+            job.stop()
+
+    def load(self) -> 'EventLoop':
+        import_subclasses()
+
+        i = 0
+        for job in BasePeriodicJob._subclasses:
```

Comment:
Enumerate

Reply:
Not sure what you mean. The `.items()` thing? I can do that instead, but the counter was an afterthought and wasn't sure the syntax order of `.items()` tuple (is it key, value or is it value, key?)

Reply:
You can use `for index, item in enumerate(list): ...`, see https://realpython.com/python-enumerate/#using-pythons-enumerate

Reply:
Okay, so looking at enumerate, how would it help with counting the number of loaded classes once outside of the loop? We're using `i` after the loop is finished, not in the loop.

Reply:
Sorry, I only looked at the annotated code here.

Can you use `len` in this case and remove `i` here entirely? Not sure if it works for this object